core/blockchain.go

```diff
@@ -842,9 +866,13 @@ func (bc *BlockChain) WriteBlockAndState(block *types.Block, receipts []*types.R
 	} else {
 		status = SideStatTy
 	}
+
+	bc.gc.LockWrite()
 	if err := batch.Write(); err != nil {
+		bc.gc.UnlockWrite()
 		return NonStatTy, err
 	}
+	bc.gc.UnlockWrite()
```

Comment:
This is a _bit_ playing with fire; holding on to one mutex (bc.mu) while obtaining another mutex. Could lead to race conditions. 

For example, the `bc.mu.RLock()` is called in `CurrentBlock`, which is called from `BackgroundGC`, which is a separate thread.

Reply:
You mean deadlock? It cannot cause a deadlock because the only thing we do under this lock is write or delete state data. Chain mutex is never used while holding this one. Still, I know that using such a lock in Blockchain is critical, but so is the GC. If we keep using it like this, we should document it very well what it does and why it does that. We should somehow avoid deleting trie nodes that reappeared just when GC removed the old entry. I am open to other suggestions though.

Note: my original proposal had an inherently safe db structure:
https://github.com/zsfelfoldi/ethereum-docs/blob/master/geth/gc_proposal.md
Unfortunately this also means that we don't know the exact key when reading and we would need a db iterator for reading every trie node. So far I could not find a db format that is both concurrency-safe and provides good performance. Using this mutex and not starting a GC while processing a block ensures that they usually don't collide and if they do anyway, it won't cause any trouble. From a practical point of view this seems to be a good solution to me if it is properly documented.